File: Sources/Scratchpad/Overlay/DisplayIdentifier.swift
import AppKit
import CoreGraphics

/// Identifies a physical display, used to key per-screen overlay windows,
/// drawing stores, and persisted drawing files.
///
/// Wraps `CGDirectDisplayID` rather than `NSScreen` itself, since AppKit
/// recreates `NSScreen` instances on every screen-configuration change while
/// the underlying display ID is stable for the life of a connected display.
struct DisplayIdentifier: Codable, Hashable {
    var rawValue: CGDirectDisplayID

    init(rawValue: CGDirectDisplayID) {
        self.rawValue = rawValue
    }

    /// A filesystem-safe representation, for naming per-display files.
    var filenameComponent: String {
        String(rawValue)
    }
}

extension DisplayIdentifier {
    /// Returns `nil` if the screen does not expose a display ID, which does
    /// not happen for any real hardware but keeps this call site honest.
    init?(screen: NSScreen) {
        guard
            let number = screen.deviceDescription[NSDeviceDescriptionKey("NSScreenNumber")] as? NSNumber
        else {
            return nil
        }
        self.init(rawValue: CGDirectDisplayID(number.uint32Value))
    }
}

File: Windows/Scratchpad/Overlay/DisplayIdentifier.cs
using System.Text.RegularExpressions;
using System.Windows.Forms;

namespace Scratchpad.Overlay;

/// Identifies a physical display, used to key per-screen overlay windows,
/// drawing stores, and persisted drawing files.
///
/// Wraps `Screen.DeviceName` (e.g. `\\.\DISPLAY1`) rather than `Screen`
/// itself, since .NET recreates `Screen` instances on every screen
/// configuration change while the device name is stable for the life of a
/// connected display -- mirrors the macOS app's `DisplayIdentifier`, which
/// wraps `CGDirectDisplayID` for the same reason. Like `CGDirectDisplayID`,
/// this name is not guaranteed stable across every hardware reconfiguration
/// (e.g. some docking-station changes can renumber `\\.\DISPLAYn`); accepted
/// as a known limitation, same as the macOS side.
public readonly record struct DisplayIdentifier(string RawValue)
{
    private static readonly Regex UnsafeFilenameCharacters = new(@"[^A-Za-z0-9_-]", RegexOptions.Compiled);

    /// A filesystem-safe representation, for naming per-display files.
    public string FilenameComponent => UnsafeFilenameCharacters.Replace(RawValue, "_");

    public static DisplayIdentifier? FromScreen(Screen screen) =>
        string.IsNullOrEmpty(screen.DeviceName) ? null : new DisplayIdentifier(screen.DeviceName);
}
